### ISSUE TITLE
Add session management method to extract auth token from the response body

### ISSUE THREAD

**Issue by kirksl** [zaproxy/zaproxy#4662]
**Describe the bug**
ZAP session management is failing to extract auth token from response body and utilize in downstream requests.

Please see the following user group discussion for details and screenshots:
https://groups.google.com/forum/#!topic/zaproxy-users/klDn2eIEiHE

**To Reproduce**
1. Startup a weekly release of ZAP in desktop mode.  We're using D-2018-05-08.
2. Open Firefox and setup it's proxy to go through ZAP.
3. Navigate to a website that requires a login, sends the login request via a POST with a JSON payload in the body(not sure if this is part of the bug) and finally returns a token in JSON format in the POST response.
4. Login to the target website.
5. ZAP should show the POST request/response under the Sites tree and successfully accumulate the traffic of the site under Sites.
6. Next.  Setup JSON based authentication so that we can Spider as a User:
6a. Right click all Sites pertaining to the target site and select "Include in context" -> "New context" and close/save dialog.
6b. Right click the POST request for the login and select "Flag as context" -> "JSON-based Auth Login Request" for the context setup in 6a, set "Username parameter" and "Password parameter" to correct JSON property from the POST request and close/save the dialog.
6c. Click POST request for the login and goto the "Response" tab.  Highlight a section of text that indicates login succeeded and select "Flag as context" -> "Authentication Logged-in indicator" and update the logged in regex pattern as needed with surrounding .* (eg \Q.*roleTokens.*\E) and close/save the dialog.
6d. Click Users tab in dialog and setup a user, username and password that can login to the target site.
6e. Click Session Management tab in dialog and set it according to how the site manages sessions and close/save dialog. (We tried both options)
7. Next select all traffic under each site if more than one under the Sites tree for the target site and delete it (We did this to see new traffic coming in to compare and see the new POST generated)
8. Finally right click the site in the Sites tree, select "Attack" -> "Spider", select a user from the "User" dropdown and click "Start Scan" and let the scan complete.

**Expected behavior**
We expected the scan to produce the same traffic we witnessed when we manually proxied through Firefox(eg the POST request/response under the Sites tree, and in this case login traffic showing appraisals, etc).  All we get is robots.txt and sitemap.xml all showing responses of "Authentication Error"

**Screenshots**
Screenshots can be found in the user group discussion:
https://groups.google.com/forum/#!topic/zaproxy-users/klDn2eIEiHE

**Software versions**
We are trying the weekly release of ZAP D-2018-05-08. (not sure if it is specific to that version or JSON)
We are using Windows 10, Firefox 59.0.3 (64-bit) 
We are using Java
C:\>java -version
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

**Errors from the zap.log file**
https://1drv.ms/u/s!AqMM4kKOXPyfinfYMfE0XPZ8XUcc

**Additional context**
NA

**Comment by kirksl** [comment]
@thc202 @psiinon @kingthorin is it possible to add a 'blocked' label to expedite or post workaround notes to this issue.  right now we have no way to use ZAP and are approaching release dates for our products.  Appreciate any help.

**Comment by kingthorin** [comment]
Unfortunately all of the devs are volunteers. They decide what to work on and when time permits. We’re always happy to welcome new contributors.

As an alternative it might be possible for you to record a zest script for logins or otherwise script the login necessities.


https://www.coveros.com/scripting-authenticated-login-within-zap-vulnerability-scanner/

https://github.com/zaproxy/community-scripts/tree/master/authentication

**Comment by kingthorin** [comment]
@kirksl can you confirm that in just proxied traffic (when you’re using the app and proxyin through zap) one of the tokens in the auth reponse does become the bearer token in the authorization header?

**Comment by kirksl** [comment]
@kingthorin confirmed.  token handed down in the login response is being used in the authorization header for downstream requests.
https://pasteboard.co/HkRpRlg.png
https://pasteboard.co/HkRq1Wo.png

if you can let us know based on this what our options are(the ones you were just mentioning like zest, etc) to get unblocked we would greatly appreciate it.

**Comment by kingthorin** [comment]
You could do something like the script in this issue: https://github.com/zaproxy/zaproxy/issues/2614

On responses check for json (old value/new value, URL, etc) parse for the token, set it into a global var. Then on request set that global value into the auth header.

**Comment by kirksl** [comment]
@kingthorin thank you for all of the links and suggestions.

i tried a zest script and it works perfect during playback.
https://pasteboard.co/Hl5pieo.png
https://pasteboard.co/Hl5o74P.png

unfortunately when spidering new pages coming in not part of the zest script failed because they were not associated with the global variable setup to hold the auth token.

next i tried to additionally add in an httpsender script.  as soon as i added/enabled it spidering failed as shown in this screenshot.  the login POST request and really no traffic came in under Sites.
https://pasteboard.co/Hl5ojcL.png

any thoughts on any likely suspects, things to try?  is it ok that i'm using a zest script in conjunction with an httpsender script.  it doesn't seem right that i'm setting a global variable "token" in the zest script and again setting that same variable "token" in the httpsender script.

thanks again for the assistance.

**Comment by rezen** [comment]
@kirksl if you need still need help I'd be willing to do a screenshare session with you. I did a POC for a similar situation with Juice Shop, below are the scripts I used to make it work.

https://github.com/rezen/zap-scripts/blob/master/authentication/auth-api.js
https://github.com/rezen/zap-scripts/blob/master/httpsender/juice-shop-maintain-auth.js

**Comment by kingthorin** [comment]
@kirksl Did you get this sorted out?

**Comment by thc202** [comment]
Title updated to be more clear about what we want with this issue.

**Comment by kingthorin** [comment]
Addressed in https://github.com/zaproxy/zaproxy/issues/5630
